scripts/kubestellar-snapshot.sh

```diff
@@ -0,0 +1,579 @@
+
+if [[ -z "${contexts[@]}" ]] ; then
+    echoerr "No context(s) found in the kubeconfig file $KUBECONFIG!"
+    exit 1
+fi
+
+echov "Validating contexts(s): "
+valid_contexts=()
+for context in "${contexts[@]}" ; do # for all contexts
+    [[ -z "$context" ]] && continue
+    if kubectl --context $context get secrets -A > /dev/null 2>&1 ; then
+        echov -e "${COLOR_GREEN}\xE2\x9C\x94${COLOR_NONE} ${COLOR_INFO}${context}${COLOR_NONE}"
+        valid_contexts+=($context)
+    else
+        echov -e "${COLOR_RED}X${COLOR_NONE} ${COLOR_INFO}${context}${COLOR_NONE}"
+    fi
+done
+
+if [[ -z "${valid_contexts[@]}" ]] ; then
+    echoerr "No valid context(s) found in the kubeconfig file $KUBECONFIG!"
+    exit 1
+fi
+
+
+###############################################################################
+# Look for KubeStellar Helm chart
+###############################################################################
+echotitle "KubeStellar:"
+helm_context=""
+for context in "${valid_contexts[@]}" ; do
+    HELM_LIST="$TMPFOLDER/helm-list.txt"
+    helm --kube-context $context list --no-headers -A 2> /dev/null > "$HELM_LIST" || true
+    while IFS= read -r line; do
+        [[ -z "$line" ]] && continue
+        name="$(echo "$line" | awk '{print $1}')"
+        namespace="$(echo "$line" | awk '{print $2}')"
+        version="$(echo "$line" | awk '{print $NF}')"
+        secret="$(kubectl --context $context get secret -n "$namespace" -l "owner=helm,name=$name" --no-headers -o name 2> /dev/null || true)"
+        secret="${secret##*/}"
+        if [[ ! -z "$(kubectl --context $context get secret $secret -n "$namespace" -o jsonpath='{.data.release}' | base64 -d | base64 -d | gzip -d | jq -r '.chart' | grep -i "KubeStellar" 2> /dev/null || true)" ]] ; then
+            helm_context="$context"
+            echo -e "- Helm chart ${COLOR_INFO}${name}${COLOR_NONE} (${COLOR_INFO}v${version}${COLOR_NONE}) in namespace ${COLOR_INFO}${namespace}${COLOR_NONE} in context ${COLOR_INFO}${context}${COLOR_NONE}"
+            echo -e "  - Secret=${COLOR_INFO}${secret}${COLOR_NONE} in namespace ${COLOR_INFO}${namespace}${COLOR_NONE}"
+            if [[ "$arg_yaml" == "true" ]] ; then
+                mkdir -p "$OUTPUT_FOLDER/kubestellar-core-chart"
+                kubectl --context $helm_context get secret $secret -o jsonpath='{.data.release}' | base64 -d | base64 -d | gzip -d | jq -r '.info'      > "$OUTPUT_FOLDER/kubestellar-core-chart/info.json"
+                kubectl --context $helm_context get secret $secret -o jsonpath='{.data.release}' | base64 -d | base64 -d | gzip -d | jq -r '.chart'     > "$OUTPUT_FOLDER/kubestellar-core-chart/chart.json"
+                kubectl --context $helm_context get secret $secret -o jsonpath='{.data.release}' | base64 -d | base64 -d | gzip -d | jq -r '.manifest'  > "$OUTPUT_FOLDER/kubestellar-core-chart/manifest.yaml"
+            fi
+            break
+        fi
+    done < "$HELM_LIST"
+    [[ ! -z "$helm_context" ]] && break
```

Comment:
This seems to be assuming that the user can not have two KubeStellar cores.

Reply:
KS require several cluster level non-namespaced resources such as KubeFlex, PCHs, CP CRDs, etc. So installing a second KS chart may fail or would create conflicts and unpredictable behaviors.. The only supported way to installa a "second" chart is to add ITSes and WDSes by disabling the installation of cluster resources such as KubeFlex, PCHs. In this scenario, the script would still work by enumerating the total resources because both chart would be in the same context. Alas the second chart would not be clearly listed in the KS section.
Based on the above considerations, I suggest to leave as is.

Reply:
What is the scenario calling for multiple KS cores? Is the idea that multiple KS cores could be sharing OCM hubs? Given a KS scheduler that must keep track of used and available WEC resources, two KS cores-schedulers would have to synchronize.

Reply:
@francostellari , @eaepstein: remember kubeconfig "contexts". Maybe you do not have as many as I do. I have a lot. Roughly, I have one for every cluster or space that I have ever worked with. Different contexts can correspond to completely unrelated and different things. While a beginning user might only kick the tires once, someone developing KubeStellar or a system incorporating KubeStellar is not unlikely to work with multiple systems. For example, I have worked no only with the kind/k3d demo systems that we have automated but also multiple systems involving our ETE clusters. I have indeed simultaenously had a KubeStellar system in those two domains.

Reply:
@MikeSpreitzer I understand your concern but the idea here is to take a snapshot of "a" KS deployment among the many that you may have across different systems. It's up to the user to provide a meaningful kubeconfig/context to the script in that case.  All data reported is only meaningful for a specific helm deployment. The script would have to add another loop to run itself on every KS helm deployments that it finds.

An alternative would be to list all of the Helm charts that it finds everywhere and then as the user to choose if more than one. To me it makes more sense that the user chooses by passing the --kubeconfig/--context

Reply:
Perhaps one thing where the script fall short in the effort to help the user when running the script in cluster1/cluster2 context by mistake is that it does not respect the current user context. So it should check the current context first before looking into other contexts.

Reply:
OK, we can start with a script that aims to work on only one KubeStellar system. We may want future work that deals with all the ones found in a given kubeconfig. A start would be to just get a list of all the contexts that have a KubeStellar core installed in them.

Reply:
I have changed the script, hopefully to show all the charts, but it will only use the first one